src/UIComponents.cpp

```diff
@@ -0,0 +1,91 @@
+#include "UIComponents.hpp"
+
+UIEntity::~UIEntity()
+{
+  entity->detach((IObserver*)this);
+};
+
+UIEntity::UIEntity(std::shared_ptr<HAEntity> _entity, lv_obj_t* _parent)
+{
+  parentContainer = _parent;
+  entity = _entity;
+  entity->attach((IObserver*)this);
+};
+
+UIButton::UIButton(std::shared_ptr<HAEntity> _entity, lv_obj_t* _parent) :
+  UIEntity(_entity, _parent)
+{
+  // START BUTTON EXAMPLE
+  btn = lv_btn_create(_parent); /*Add a button the current screen*/
+  lv_obj_set_size(btn, 240, 50);
+  lv_obj_center(btn);
+  lv_obj_add_flag(btn, LV_OBJ_FLAG_CHECKABLE);
+  lv_obj_add_event_cb(btn, UIButton::btn_press_cb, LV_EVENT_ALL, reinterpret_cast<void*>(&entity)); /*Assign a callback to the button*/
+
+  lv_obj_t* label = lv_label_create(btn); /*Add a label to the button*/
+  lv_label_set_text(label, _entity->name.c_str()); /*Set the labels text*/
+  lv_obj_center(label);
+  uiupdate();
+};
+
+void UIButton::uiupdate()
+{
+  auto state = entity->getJsonState();
+  if (state["state"] == "on") {
+    lv_obj_add_state(btn, LV_STATE_CHECKED);
+  }
+  else {
+    lv_obj_clear_state(btn, LV_STATE_CHECKED);
+  }
+};
+
+void UIButton::btn_press_cb(lv_event_t* e)
+{
+  lv_event_code_t code = lv_event_get_code(e);
+
+  std::shared_ptr<HAEntity> ent = *reinterpret_cast<std::shared_ptr<HAEntity>*>(e->user_data);
+  if (code == LV_EVENT_VALUE_CHANGED) {
```

Comment:
I'm slightly surprised this is the event value for clicking a button. Feels like something that might fire twice if you hold the button for a second.

I wonder if we can avoid future race condition-like trouble by sending an explicit off or on here, based on (the opposite of) the current button state. Then again, I imagine we'll stop using buttons for off/on things soon.

Reply:
Not sure, it seems the button changes on mouse-up. I don't have a touchscreen on my laptop, so not sure what that does.
It works though. Long press works fine and doesn't cause issues.

Reply:
ok great :)